pkg/api/commsRetrier.go

```
@@ -39,6 +36,85 @@ func NewCommsRetrier(maxAttempts int, retryBackoffSeconds float64, timeService s
+	maxAttempts int,
+	retryBackoffSeconds float64,
+	timeService spi.TimeService,
+	authenticator spi.Authenticator,
+) (CommsRetrier, error) {
+	var err error
+	var apiClient *galasaapi.APIClient
+	var commsRetrier CommsRetrier
+
+	apiClient, err = authenticator.GetAuthenticatedAPIClient()
+	if err == nil {
+		commsRetrier = &CommsRetrierImpl{
+			maxAttempts: maxAttempts,
+			retryBackoffSeconds: retryBackoffSeconds,
+			timeService: timeService,
+			authenticator: authenticator,
+			apiClient: apiClient,
+		}
+	}
+	return commsRetrier, err
+}
+
+// ExecuteCommandWithRetries tries to run a given execution function until we've tried enough, it worked, 
+// or it has failed too many times with rate limit or auth issues. If an unauthorized error is encountered,
+// then this function will attempt to re-authenticate with the API server.
+func (retrier *CommsRetrierImpl) ExecuteCommandWithRetries(
+	commandExecutionFunc func(apiClient *galasaapi.APIClient) error,
+) error {
+	var err error
+    isDone := false
+	attempt := 0
+	maxAttempts := retrier.maxAttempts
+	retryBackoffSeconds := retrier.retryBackoffSeconds
+	timeService := retrier.timeService
+
+    for !isDone {
+
+        err = commandExecutionFunc(retrier.apiClient)
+
+        isDone = true
+        if err != nil {
+
+			isRetryRequired := false
+
+            // Try to convert the error received from the command into an API error
+            galasaError, isGalasaError := err.(galasaErrors.GalasaCommsError)
+            if isGalasaError {
+
+				// If the command encountered an unauthorized error from the API server,
+				// attempt to log in again to get a new JWT and use that in subsequent requests
+				if galasaError.IsReauthRequired() {
+					log.Printf("Reauthentication required. Login attempt %v/%v", (attempt + 1), maxAttempts)
+					var newApiClient *galasaapi.APIClient
+					var apiClientErr error
+					newApiClient, apiClientErr = retrier.authenticator.GetAuthenticatedAPIClient()
```

If the re-auth fails then the error gets lost.